tegra: Add support for commit service and verifying root and boot slot alignment

dwalkes opened the pull request:
Tegra platforms Xavier NX and TX2 support a [redundant boot with A/B update scheme](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fbootloader_update_agx_tx2.html%23) using `nvbootctrl` as the mechanism to setup and control boot slots.  In addition, the TX2 supports either a cboot or u-boot based boot scheme, with u-boot as the default for the MACHINE.  For Xavier NX, cboot is the default bootloader.  Nano supports only u-boot based boot, and [nano boot redundancy](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fbootloader_update_nano_tx1.html%23) is not currently supported by the meta-mender-community meta-mender-tegra layer.

With the cboot based implementations, A/B slots used for bootloader directly correspond to the root filesystem slot selected at boot time.  The `nvbootctrl` slot is the single source of truth, and mender bases the selection of the active rootfs for update purposes using the fake libubootenv scripts.

With u-boot redundant boot implementation in TX2, the A/B slots used for the bootloader do not directly correspond to the root filesystem slot selected at boot time.  The u-boot parameters are controlled by mender, while the boot slots are controlled by `nvbootctrl`.  This can lead to a scenario where the NVIDIA boot components update the boot slot, based on [Update State Machine](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fbootloader_update_agx_tx2.html%23wwpID0E0SD0HA), while the mender components do not change.  This can result in a mismatch of the boot slot and root filesystem slot.  This often manifests as failure during commit as discussed at https://github.com/OE4T/meta-mender-community/issues/7 and [this post on mender hub](https://hub.mender.io/t/failing-updates-on-tegra-tx2-due-to-issues-with-nvbootctrl/2765/2).  

The easiest way to reproduce this problem on u-boot systems is by forgetting to run `mender -commit` after installing and rebooting in standalone install mode.  When you forget this step, the NVIDIA bootloader starts a retry count, waiting for `nvbootctrl mark-boot-successful` to be set as described in [bootloader documentation](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fbootloader_update_agx_tx2.html%23wwpID0E0PD0HA).  After 7 boot attempts without committing, the NVIDIA bootloader rolls back to the previous slot.  However, since mender and u-boot are not synchronized, the mender rootfs slot still references the wrong slot.  If there's no mismatch between bootloader version and rootfs version you won't actually notice this difference unless you specifically look at the output of `nvbootctrl get-current-slot`.  If you try another `mender -install` (or install through hosted mender) with a rootfs/boot slot mismatch, the artifact install step succeeds, however the `mender -commit` fails with messages like this:
```
---------- end of script output
ERRO[0000] ArtifactCommit_Enter script failed: statescript: error executing 'ArtifactCommit_Enter_80_bl-check-update': 1 : exit status 1
Rolling back Artifact...
INFO[0000] Setting partition for rollback: 33
ERRO[0001] statescript: error executing 'ArtifactCommit_Enter_80_bl-check-update': 1 : exit status 1
```
since the commit script detects a bootloader slot mismatch.  At this point the only way to recover is to manually reset the boot slot to match the root partition slot using `nvbootctrl set-active-boot-slot` or changing the u-boot environment variables and associated root filesystem slot.

To avoid this scenario, this PR proposes two changes:
* A wrapper script which detects when mender is used in standalone mode and schedules an auto-commit on the next update. 
 This makes it less likely you will end up in the scenario described above which is probably the most likely case where this will happen and is particularly important for u-boot platforms, but also useful for others.
* Checks which validate and warn about boot and root filesystem slot mismatches at boot time for u-boot platforms and automatically:
  * Set the boot slot controlled by NVIDIA software to match the rootfs slot controlled by mender, using the mender designated rootfs slot as the source of truth (see https://github.com/BoulderAI/meta-mender-community/pull/1#issuecomment-723463721).
  * Stop the mender-client service to disable upgrade attempts before the next reboot.  This prevents the possibility of corrupting the boot or root filesystem slot with a partial mender update and preventing the possiblity to get the boot slots back in alignment.

This makes the u-boot logic more closely match the cboot logic for mender update and tries to correct issues with boot slot and rootfs partition mismatch as they happen rather than waiting for the next update attempt.

kekiefer commented:
This seems reasonable. I've changed the base to master because that's what we're working on right now. We'll backport as appropriate from there once master is settled.

Commit:
Adding scripts to align boot and rootfs partition

The mender boot slot values are not in sync with nvbootctrl slot values. The values are updated correctly in the mender_boot_part variable, but the nvbootctrl fails to sync up with the latest boot partition to use on the next reboot after a mender update sometimes.
This commit would rectify the issue by identifying the mismatch and giving a Warning and dumping all the data for debug.
At the same time it would set the active boot partition based on the value in mender_boot_partition, so that there is no mismatch going forward.
This script would run after every mender update and will run as a one shot service.

dwalkes commented on meta-mender-tegra/recipes-mender/mender-client/mender-client_%.bbappend:
```diff
@@ -1 +1,22 @@
+FILES_${PN} += "${sysconfdir}/mender/mender-client-commit.sh"
+FILES_${PN} += "${bindir}/mender"
+
+SYSTEMD_SERVICE_${PN} = "mender-client-commit.service"
```
@jajoosiddhant you need to change this to += so it doesn't replace the mender client service setup [here](https://github.com/mendersoftware/meta-mender/blob/master/meta-mender-core/recipes-mender/mender-client/mender-client.inc#L30)

dwalkes replied:
Right now I see:
```
systemctl status mender-client
Unit mender.service could not be found.
● mender-client.service - Mender OTA update service
     Loaded: loaded (/lib/systemd/system/mender-client.service; disabled; vendor preset: disabled)
     Active: inactive (dead)
```

dwalkes replied:
actually I think you should use `append_mender-uboot` here

dwalkes commented on meta-mender-tegra/recipes-mender/mender-client/files/mender:
```diff
@@ -0,0 +1,19 @@
+
+echo "Starting mender wrapper"
+
+echo $@ | grep 'install\|-install' > /dev/null
```
@jajoosiddhant include spaces around the install text here to prevent match with argument content (like filenames)

dwalkes committed:
Update mender client commit support

Move verify boot slot alignment scripts into separate files so we
can support the client commit service in non redundant bootloader
cases like t210 and cboot cases where no verification is necessary

Signed-off-by: Dan Walkes <danwalkes@boulderai.com>

dwalkes committed:
Handful of Fixes

 * Use correct quoting on echo arguments so match succeeds
 * Move verify script into bin dir, rename with mender-tegra prefix

dwalkes committed:
Prevent install from mender install on boot slot mismatch

When boot slots aren't aligned (as detected/corrected by the verify
alignment script) we need to set a marker file in the volatile FS
to prevent future mender -install attempts from running until
the next reboot when boot slots are once again aligned.

madisongh commented on meta-mender-tegra/recipes-mender/mender-client/files/mender:
```diff
@@ -10,6 +10,11 @@ if [ $? -eq 0 ]; then
     # Exit with failure and error message if we don't have alignment
     # between boot slot and rootfs.  It's not safe to update in this case
     mender-tegra-verify-boot-rootfs-slot-alignment || exit 1
+    if [ -e /var/volatile/mender-tegra-boot-slot-mismatch-install-disabled ]; then
```
Suggest using `/run` instead of `/var/volatile` as the location for this sentinel file.

dwalkes replied:
Thanks for the suggestion!  @jajoosiddhant can you please make this change before you test on Monday?

jajoosiddhant replied:
Sure

jajoosiddhant replied:
Verified

dwalkes committed:
Add mender test script
